lib/service/chat_service.dart

```diff
@@ -0,0 +1,213 @@
+              }
+            } else {
+              try {
+                final newMessages =
+                    (response.payload["messages"] as List<dynamic>)
+                        .map((e) => app.Message.fromJson(e))
+                        .toList();
+                for (var element in _listeners) {
+                  element.onResponseMessageReturnPayload(
+                      newMessages, response.id);
+                }
+              } catch (e) {
+                log.info("[CHAT page] RESP error: $e");
+              }
+            }
+            break;
+          default:
+            break;
+        }
+      },
+      onDone: () async {
+        log.info("[CHAT] onDone");
+
+        Future.delayed(const Duration(seconds: 5), () async {
+          if (_address != null && _id != null && _wallet != null) {
+            _address = null;
+            _id = null;
+            _wallet = null;
+            log.info("[CHAT] _websocketChannel reconnecting");
+            await connect(
+              address: _address!,
+              id: _id!,
+              wallet: _wallet!,
+            );
+            for (var element in _listeners) {
+              element.onDoneCalled.call();
+            }
+          }
+        });
+      },
+    );
+  }
+
+  @override
+  void addListener(ChatListener listener) {
+    _listeners.add(listener);
+  }
+
+  @override
+  Future<void> dispose() async {
+    log.info("[CHAT] disconnect");
+    _address = null;
+    _id = null;
+    _wallet = null;
+    await _websocketChannel?.sink.close();
+    _websocketChannel = null;
+  }
+
+  @override
+  void sendMessage(dynamic message) {
+    _websocketChannel?.sink.add(message);
+  }
+
+  Future<Map<String, dynamic>> _getHeader(
+      String link, Pair<WalletStorage, int> wallet, String address) async {
+    final Map<String, dynamic> header = {};
+    final timestamp = DateTime.now().millisecondsSinceEpoch;
+    header["X-Api-Timestamp"] = timestamp;
+    final canonicalString = List<String>.of([
+      link,
+      "",
+      timestamp.toString(),
+    ]).join("|");
+    final hmacSha256 = Hmac(sha256, utf8.encode(Environment.chatServerHmacKey));
+    final digest = hmacSha256.convert(utf8.encode(canonicalString));
+    final sig = bytesToHex(digest.bytes);
+    header["X-Api-Signature"] = sig;
+
+    final pubKey = await wallet.first.getTezosPublicKey(index: wallet.second);
+    final authSig = await injector<TezosService>().signMessage(wallet.first,
+        wallet.second, Uint8List.fromList(utf8.encode(timestamp.toString())));
+    final token = await injector<ChatAuthService>().getAuthToken({
+      "address": address,
+      "public_key": pubKey,
+      "signature": authSig,
+      "timestamp": timestamp
+    });
+    header["Authorization"] = "Bearer $token";
+    return header;
+  }
+
+  @override
+  void removeListener(ChatListener listener) {
+    _listeners.remove(listener);
```

Comment:
consider checking the `listener` exists in the list to avoid unexpected crashes if accidentally pass a wrong object.

Reply:
i don't thinks that is the case. it returns false if we pass wrong object
/// The method has no effect if [value] was not in the list.
  /// ```dart
  /// final parts = <String>['shoulders', 'knees', 'toes'];
  /// // Note: 'head' has already been removed.
  /// final retVal = parts.remove('head'); // false
  /// print(parts); // [shoulders, knees, toes]
  /// ```
  bool remove(Object? value);